### ISSUE TITLE
std.Build.Step.Test.setExecCmd no longer does anything

### ISSUE THREAD

**Issue by gcoakes** [ziglang/zig#17756]
### Zig Version

0.12.0-dev.1264+d1230842a

### Steps to Reproduce and Observed Behavior

1. `zig init-lib`
2. Apply the following change:

```diff
diff --git a/build.zig b/build.zig
index cacab1f..db1abfe 100644
--- a/build.zig
+++ b/build.zig
@@ -36,6 +36,7 @@ pub fn build(b: *std.Build) void {
         .target = target,
         .optimize = optimize,
     });
+    main_tests.setExecCmd(&[_]?[]const u8{ "sh", "-c", "echo worked > did-it.txt", null });
 
     const run_main_tests = b.addRunArtifact(main_tests);
 

```

3. Observe `did-it.txt` was not created implying `sh` was never run.

This was observed while trying to add a coverage flag to my `build.zig`. So, in my case, it would have looked more like:

```diff
diff --git a/build.zig b/build.zig
index cacab1f..3ddac03 100644
--- a/build.zig
+++ b/build.zig
@@ -36,6 +36,7 @@ pub fn build(b: *std.Build) void {
         .target = target,
         .optimize = optimize,
     });
+    main_tests.setExecCmd(&[_]?[]const u8{ "kcov", "kcov-out", null });
 
     const run_main_tests = b.addRunArtifact(main_tests);
 

```

This behavior is only observed when `--listen=-` is present in the arguments for `zig test`. I can copy the arguments from `zig build test --verbose` and run them without `--listen=-`. In that case, it works:

```sh
$ zig test /home/gcoakes/src/repro/src/main.zig --cache-dir /home/gcoakes/src/repro/zig-cache --global-cache-dir /home/gcoakes/.cache/zig --name test --test-cmd kcov --test-cmd kcov-out --test-cmd-bin
All 1 tests passed.
$ ls
build.zig  kcov-out  LICENSE  README.md  src  zig-cache  zig-out
```

### Expected Behavior

The test executable to be run using the specified command/arguments.

**Comment by andrewrk** [comment]
It's not really intended for users to pass `--listen=-` - this is for a parent process to pass when it wants to communicate over a binary channel with the child process, and take over the responsibility of executing the test via some command.

**Comment by xdBronch** [comment]
this isnt about passing it manually, `--listen=-` is passed by the build system so `--test-cmd` (or more like the build system equivalent) does nothing

**Comment by gcoakes** [comment]
@xdBronch, correct. I frankly don't want the whole RPC being used if I could help it (I would disable the TTY line erasing also if I could find out how). I even added this to my tests, and it's still using the RPC:

```zig
    unit_tests.test_server_mode = false;
```

**Comment by andrewrk** [comment]
Ah, `setExecCmd`. I got confused by the title, sorry.

**Comment by andrewrk** [comment]
I'd appreciate a proper explanation of your problem with the communication channel between the build runner and the test runner.

As for progress output, I agree a "quiet" flag would be nice to have, although I don't understand the problem, since it only shows up when stderr is a tty. You don't want to know the progress of the build?

**Comment by gcoakes** [comment]
<details><summary>@andrewrk, I just preferred the old output from <code>zig build test</code> where the names of the tests and their results would spew out to the terminal. That was a tangent from the core issue. Sorry for that.</summary> i.e.:<br/><pre>$ zig build test 2>&1 | cat
1/1 test.hello... 1/5 test_0... OK
2/5 test.request parsing... OK
All 1 tests passed.
OK
3/5 test.response writing... OK
4/5 test.fuzz chunked round trip... OK
5/5 decltest.stringToEnumIgnoreCase... OK
All 5 tests passed.</pre> I can get it to act how I want with some bash trickery (<code>zig build test 2>&1 | cat</code>) and changes to my build.zig. I just figured it was a meaningless personal preference.</details>

The core issue is that `setExecCmd` doesn't seem to get applied in some circumstances. Aside from my mess of a build.zig, the issue can be minimally reproduced with `zig init-lib` and the diff from the first comment.

**Comment by andrewrk** [comment]
> I just preferred the old output from zig build test where the names of the tests and their results would spew out to the terminal.

the build runner is a multiplexer. it concurrently runs multiple test suites, so they can't print to stderr or they would clobber each other.

**Comment by gcoakes** [comment]
Thank you! I didn't know about `--summary all`. That's much better than what I was doing. Before I would run `zig build test`, and I would just end up with a new terminal line. i.e.:

```console
$ zig build test
$
```

It was really frustrating because I wanted to see positive confirmation that something actually ran.

**Comment by xdBronch** [comment]
btw i have run into this problem as well. i was trying to teach someone how you could debug zig tests and `--test-cmd` with gdb worked great but when i tried to apply it to the build script nothing happened. maybe it'd be an anti-pattern for something like gdb to run in the middle of a `zig build test` run but its always nice to have everything contained in a single build.zig

**Comment by andrewrk** [comment]
@xdBronch that's a good use case, let's try to improve things to make that handy 👍

**Comment by gcoakes** [comment]
I think I found something interesting. It looks like the server code exits the process immediately following the serve loop: https://github.com/ziglang/zig/blob/13c7aa5fefef9b1338951cf0d01c04345201d996/src/main.zig#L3583-L3594

Without `--listen=-`, the `zig test` process would normally enter into this area where the arguments are properly applied: https://github.com/ziglang/zig/blob/13c7aa5fefef9b1338951cf0d01c04345201d996/src/main.zig#L3675-L3694

I used `setExecCmd(&[_]?[]const u8{ "echo", null })` and strace'd the whole build process (`strace -y -o strace.log --follow-forks --output-separately -s 8192 zig build test-unit`). Using that I could observe the client-server interactions in the read/write system calls.
<details><summary>Interpreting those messages, it looks like the client sends an <code>.update</code> and <code>.exit</code> message. Then, the server -- which is <code>zig test --test-cmd ...</code> -- responds with <code>.zig_version</code>, several <code>.progress</code>, and one <code>.emit_bin_path</code>. No further other interactions are observed along that pipe. The following is from the perspective of the client (build_runner.zig if I'm not mistaken):</summary>
<pre>
write(10<pipe:[375465]>, "\1\0\0\0\0\0\0\0", 8) = 8
write(10<pipe:[375465]>, "\0\0\0\0\0\0\0\0", 8) = 8
poll([{fd=11<pipe:[413548]>, events=POLLIN}, {fd=13<pipe:[413549]>, events=POLLIN}], 2, 2147483647) = 1 ([{fd=11, revents=POLLIN}])
read(11<pipe:[413548]>, "\0\0\0\0\31\0\0\0000.12.0-dev.1326+9ff9ea38e", 512) = 33
poll([{fd=11<pipe:[413548]>, events=POLLIN}, {fd=13<pipe:[413549]>, events=POLLIN}], 2, 2147483647) = 1 ([{fd=11, revents=POLLIN}])
read(11<pipe:[413548]>, "\2\0\0\0\31\0\0\0Semantic Analysis [2003] ", 512) = 33
poll([{fd=11<pipe:[413548]>, events=POLLIN}, {fd=13<pipe:[413549]>, events=POLLIN}], 2, 2147483647) = 1 ([{fd=11, revents=POLLIN}])
read(11<pipe:[413548]>, "\2\0\0\0\31\0\0\0Semantic Analysis [3204] ", 512) = 33
poll([{fd=11<pipe:[413548]>, events=POLLIN}, {fd=13<pipe:[413549]>, events=POLLIN}], 2, 2147483647) = 1 ([{fd=11, revents=POLLIN}])
read(11<pipe:[413548]>, "\2\0\0\0\31\0\0\0Semantic Analysis [3784] ", 512) = 33
poll([{fd=11<pipe:[413548]>, events=POLLIN}, {fd=13<pipe:[413549]>, events=POLLIN}], 2, 2147483647) = 1 ([{fd=11, revents=POLLIN}])
read(11<pipe:[413548]>, "\2\0\0\0\20\0\0\0LLVM Emit Object", 512) = 24
poll([{fd=11<pipe:[413548]>, events=POLLIN}, {fd=13<pipe:[413549]>, events=POLLIN}], 2, 2147483647) = 1 ([{fd=11, revents=POLLIN}])
read(11<pipe:[413548]>, "\2\0\0\0\20\0\0\0LLVM Emit Object", 512) = 24
poll([{fd=11<pipe:[413548]>, events=POLLIN}, {fd=13<pipe:[413549]>, events=POLLIN}], 2, 2147483647) = 1 ([{fd=11, revents=POLLIN}])
read(11<pipe:[413548]>, "\2\0\0\0\20\0\0\0LLVM Emit Object", 512) = 24
poll([{fd=11<pipe:[413548]>, events=POLLIN}, {fd=13<pipe:[413549]>, events=POLLIN}], 2, 2147483647) = 1 ([{fd=11, revents=POLLIN}])
read(11<pipe:[413548]>, "\2\0\0\0\20\0\0\0LLVM Emit Object", 512) = 24
poll([{fd=11<pipe:[413548]>, events=POLLIN}, {fd=13<pipe:[413549]>, events=POLLIN}], 2, 2147483647) = 1 ([{fd=11, revents=POLLIN}])
read(11<pipe:[413548]>, "\2\0\0\0\20\0\0\0LLVM Emit Object", 512) = 24
poll([{fd=11<pipe:[413548]>, events=POLLIN}, {fd=13<pipe:[413549]>, events=POLLIN}], 2, 2147483647) = 1 ([{fd=11, revents=POLLIN}])
read(11<pipe:[413548]>, "\2\0\0\0\20\0\0\0LLVM Emit Object", 512) = 24
poll([{fd=11<pipe:[413548]>, events=POLLIN}, {fd=13<pipe:[413549]>, events=POLLIN}], 2, 2147483647) = 1 ([{fd=11, revents=POLLIN}])
read(11<pipe:[413548]>, "\2\0\0\0\10\0\0\0LLD Link", 512) = 16
poll([{fd=11<pipe:[413548]>, events=POLLIN}, {fd=13<pipe:[413549]>, events=POLLIN}], 2, 2147483647) = 1 ([{fd=11, revents=POLLIN}])
read(11<pipe:[413548]>, "\3\0\0\0I\0\0\0\0/home/gcoakes/src/watt/zig-cache/o/2f71fa0057d09110da90fe0d7c559f6b/test", 512) = 81
</pre></details>

Basically, the process which has knowledge of the exec arguments doesn't actually run the test. It just reports the path to the binary, then exits. The process that actually forks into the test process is the client from above. Looking at that code (`std.zig.Build.Step.Run.evalZigTest`), it seems like it pulls it's argv from the run step, not the compile step. This is very obvious in retrospect, but then that begs the question: why is `setExecCmd` an option on the compilation step? Judging by the TODO in the server code, is it just not implemented yet? I think the server change looks somewhat simple, but I assume that entails much larger changes to `std.zig.Build.Step.Run` and `std.zig.Build.Step.Compile`. Am I correct on that?

**Comment by Snektron** [comment]
I'm running into this issue with the Zig SPIR-V test executor. Its a little unclear to me how exactly this serving architecture is supposed to work together with runnings tests. As I understand it, currently tests are supposed to be built and run separately (from https://ziglang.org/learn/build-system/#testing):
```zig
const test = b.addTest(...);
const run_test = b.addRunArtifact(test);
const test_step = b.step("test", "Run the tests");
test_step.dependOn(&run_test.step);
```
If this is the intended way that it will work in the future, then it seems sufficient to me to make `addRunArtifact` simply adhere the `exec_cmd_args` and `test_server_mode` from the test.

> This was observed while trying to add a coverage flag to my build.zig. So, in my case, it would have looked more like:

Note that you can always invoke the test by manually creating a run step:
```zig
const run_step = b.addSystemCommand(&.{"kcov", "kcov-out"});
run_step.addArtifactArg(my_test);
```

**Comment by davo9819** [comment]
Hello, I'm having the same problem with `std.Build.Step.Test.setExecCmd`.

Zig Version: 0.13.0

Running the tests:
```bash
zig build test -Dtest-coverage --verbose
```
Output:
```bash
/home/david/.local/zig/zig test -ODebug -Mroot=/home/david/PROJECTS/coverage/src/main.zig --cache-dir /home/david/PROJECTS/coverage/.zig-cache --global-cache-dir /home/david/.cache/zig --name test --test-cmd kcov --test-cmd kcov-output --test-cmd-bin --listen=-
```
But no coverage report is generated. Manually calling `zig test ...` (minus the `--listen=-`) gives the desired output, tests are passed and the coverage report is generated.

**Comment by xdBronch** [comment]
the fix for this was merged after 0.13 was released
